NDTensors/src/dense/dense.jl

```diff
@@ -3,73 +3,145 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::DataT) where {ElT,DataT<:AbstractArray}
+    return new{ElT,DataT}(data)
   end
+end
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  return Dense{ElT,typedDataT}(typedDataT())
+end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
 
-  # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(dim::Integer) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim))
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
+  end
+  return Dense{ElT,typedDataT}(similar(typedDataT, dim(inds)))
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT<:AbstractArray}
+  if eltype(data) != ElR
+    data = similartype(typeof(data), ElR)(data)
+  end
+  return Dense{ElR,DataT}(data)
 end
 
-function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+function Dense{DataT}() where {DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, default_eltype())
+  ElT = eltype(typedDataT)
+  return Dense{ElT,typedDataT}()
```

Comment:
This seems strange to me, I don't think we want to ever treat the first type parameter of `Dense` as the data type, only as the element type.

Reply:
Also we shouldn't be changing the `DataT` if it is specified. What's the purpose of this function?

Reply:
If you want to define versions where the data type is specified but the element type isn't, the way to do that is:
```julia
function Dense{<:Any,DataT}()
  return Dense{eltype(DataT),DataT}()
end
```
which would be called like `Dense{<:Any,Vector{Float32})()`. It's a bit verbose but the best we have since type parameters are positional.

Reply:
@mtfishman In this function I don't actually change the datatype. I don't know if this is something that would happen often but my thought was that you might want to run a program completely with a `default_eltype` but change the datatype say from `Vector` to `DistibutedVector`. Then you could construct with `Dense{ArrayType}()` where Arraytype is specified. 
Also this code does not actually change the datatype. What it does is cover the case if you call `Dense{Vector}` vs `Dense{Vector{Float64}}`. I updated the `set_eltype_if_unspecified` to take an `AbstractArray`. It does nothing if `eltype` is specified or sets the eltype to the a provide type if unspecified (in general this is `default_eltype()`). This way we don't need to have `{DataT<:AbstractArray{ElT}} where {ElT}`

Reply:
There are two issues:

1. If you call `Dense{T}([...])`, `T` should never be interpreted as the `datatype`, only as the `eltype`. We should consider this a strict rule about constructor calls. For specifying the `datatype` and not the `eltype`, we should use the notation `Dense{<:Any,DataT}([...])`. I would argue that this should only ever return an object of type `Dense{eltype(DataT),DataT}` (maybe in some cases it is better to infer the element type from an argument input but in general I think `eltype(DataT)` is best).
2. If the behavior is that `Dense{<:Any,Vector}()` returns an object of type `Dense{<:Any,Vector{Float64}}`, I consider that changing the data type, since `datatype(Dense{<:Any,T}()) != T` (I would take a strict definition that the data type is the result of calling `datatype`, so you can't ignore the element type).

As a principle (I would say a rare case when there is a very strict rule to follow), type parameter inputs of constructors should be taken very literally, unless they are completely unspecified. So for any type, we should have constructors that follow the pattern:
```julia
struct T{A,B}
  # Some fields
end
T{A,B}(args...) = # Must return an object of type `T{A,B}`
T(args...) = T{infer_A(args...),infer_B(args...)}(args...)
T{A}(args...) = T{A,infer_B(A, args...)}(args...)
T{<:Any,B}(args...) = T{infer_A(B, args...),B}(args...)
```
For behavior that is less strict about type inputs, we should rely on functions that perform conversions (such as `set_eltype` or `adapt`) or passing types as arguments, not as type parameters.

Reply:
1. I can modify the the `Dense{DataType}` constructors to match these principles.
However, I just want to point out that from my constructor tree for dense it is not possible for `Dense{T}([...])` to infer `T` as the `datatype` because there is no `Dense{DataT<:AbstractArray}(data::AbstractArray)` constructor, only a `Dense{ElT}(data::AbstractArray) where {ElT}` constructor. This is currently some functionality of the code as its written
```
julia> NDTensors.Dense{Vector{Float32}}()
0-element NDTensors.Dense{Float32, Vector{Float32}}
julia> NDTensors.Dense{Vector{Float32}}([1,2,3])
ERROR: MethodError: Cannot `convert` an object of type Int64 to an object of type Vector{Float32}
julia> NDTensors.Dense{Vector{Float32}}([[1,2,3],[2],[3,4,3]])
typeof(d) = Vector{Vector{Float32}}
3-element NDTensors.Dense{Vector{Float32}, Vector{Vector{Float32}}}:
 [1.0, 2.0, 3.0]
 [2.0]
 [3.0, 4.0, 3.0]
 ``` 
 Any time actual data is provided to `Dense` it will consider it the eltype because `ElT` is general and not required to be a number. 
 
 2.
 The behavior of `Dense{<:Any, Vector}()` would return an object of `Dense{eltype(Dense), Vector{eltype(Dense)}`. What I am trying to do in these constructors is be more stringent to ensure that `ElT = eltype(DataT) = eltype(data)` and `typeof(data) = DataT`. There is an example in the ITensors tests where previously the code constructed a dense via `Dense{ComplexF64, Vector{ComplexF64})(Vector{Float64}([...])`. Also I found that we can do something like this still 
 ```
 julia> NDTensors.Dense{Float32, Vector{Float64}}(Vector{Float64}([2,3,4]))
3-element NDTensors.Dense{Float32, Vector{Float64}}:
 2.0
 3.0
 4.0
 ``` 
I know that these cases aren't "normal behavior" for us or anyone using ITensors but I just to ensure that it is not possible to construct such ambiguous objects because they could cause problems down the line.

Reply:
Right, I can see that in principle there are cases when we can unambiguously decide that for a call `Dense{T}(args...)`, `T` _could_ be interpreted as the `datatype`, but it is a question of what we _should_ do, and I think it is playing way to fast-and-loose with constructor calls to ever interpret `T` (the first type parameter) as the `datatype`.

I'm confused by you statement that `Dense{<:Any, Vector}()` should return an object of type `Dense{eltype(Dense),Vector{eltype(Dense)}`, I'm not really sure what that means. I would vote for a very general definition:
```julia
Dense{<:Any,DataT}(args...) where {DataT} = Dense{eltype(DataT},DataT}(args...)
```
in which case we always have the property:
```julia
datatype(Dense{<:Any,T}(args...)) == T
eltype(Dense{<:Any,T}(args...)) == eltype(T)
```
Then, by that definition, `Dense{<:Any,Vector}(args...)` returns an object of type `Dense{Any,Vector}`. Even if we could infer a more specific `eltype` than `Any` based on the `args...`, I don't think we should, and in general we should opt for taking the specified type parameters very literally and also keeping them consistent with each other.

That's helpful context that you are trying to deal with those types of corner cases where the type parameters are inconsistent with each other, and it is worth thinking about how to deal with that. I would say that for now, we just try to ensure that our own code always does the correct/safest thing (i.e. we should never generate a type `Dense{Float32,Vector{Float64}}` in our code, for example by using rules I lay out above). It seems like we could catch some of those type parameter consistency mistakes in the inner constructor, for example we could add a check:
```julia
function Dense{ElT,DataT}(data)
  @assert ElT == eltype(DataT)
  return new{ElT,DataT}(convert(DataT, data))
end
```

Reply:
One way to think about this is that we could have thought about designing the `Dense` type with just one type parameter:
```julia
struct Dense{DataT}
  data::DataT
end
```
in which case the definition of `eltype` would be determined directly from the `DataT` parameter:
```julia
datatype(::Dense{DataT}) where {DataT} = DataT
eltype(x::Dense) = eltype(datatype(x))
```
We could think of the `ElT` parameter in `Dense` as just a convenient way of working with the `eltype` of `DataT` for dispatch, constructor calls, etc. but for all intents and purposes it is really just `eltype(DataT)` by definition.